api/src/services/application-csv-export.service.ts

```diff
@@ -815,15 +698,380 @@ export class ApplicationCsvExporterService
+        user.id
+      }-${new Date().getTime()}.csv`,
+    );
+
+    await this.createLotterySheet(filename, queryParams);
+    const file = createReadStream(filename);
+    return new StreamableFile(file);
+  }
+
+  /**
+   *
+   * @param filename
+   * @param queryParams
+   * @returns generates the lottery sheet
+   */
+  async createLotterySheet<QueryParams extends ApplicationCsvQueryParams>(
+    filename: string,
+    queryParams: QueryParams,
+  ): Promise<void> {
+    const applications = await this.prisma.applications.findMany({
+      select: {
+        id: true,
+        preferences: true,
+        householdMember: {
+          select: {
+            id: true,
+          },
+        },
+        applicationLotteryPositions: {
+          select: {
+            ordinal: true,
+            multiselectQuestionId: true,
+          },
+          where: {
+            multiselectQuestionId: null,
+          },
+          orderBy: {
+            ordinal: OrderByEnum.DESC,
+          },
+        },
+      },
+      where: {
+        listingId: queryParams.listingId,
+        deletedAt: null,
+        markedAsDuplicate: false,
+      },
+    });
+
+    // get all multiselect questions for a listing to build csv headers
+    const multiSelectQuestions =
+      await this.multiselectQuestionService.findByListingId(
+        queryParams.listingId,
+      );
+
+    // get maxHouseholdMembers associated to the selected applications
+    let maxHouseholdMembers = 0;
+    applications.forEach((app) => {
+      if (app.householdMember?.length > maxHouseholdMembers) {
+        maxHouseholdMembers = app.householdMember.length;
+      }
+    });
+
+    const csvHeaders = await this.getCsvHeaders(
+      maxHouseholdMembers,
+      multiSelectQuestions,
+      queryParams.timeZone,
+      queryParams.includeDemographics,
+      true,
+    );
+
+    return this.csvExportHelper(
+      filename,
+      mapTo(Application, applications),
+      csvHeaders,
+      queryParams,
+      true,
+    );
+  }
+
+  /**
+   *
+   * @param queryParams
+   * @param req
+   * @returns generates the lottery results for a listing
+   */
+  async lotteryGenerate<QueryParams extends ApplicationCsvQueryParams>(
+    req: ExpressRequest,
+    res: Response,
+    queryParams: QueryParams,
+  ): Promise<SuccessDTO> {
+    const user = mapTo(User, req['user']);
+    await this.authorizeCSVExport(user, queryParams.listingId);
+
+    const listing = await this.prisma.listings.findUnique({
+      select: {
+        id: true,
+        lotteryLastRunAt: true,
+        lotteryStatus: true,
+      },
+      where: {
+        id: queryParams.listingId,
+      },
+    });
+    if (listing?.lotteryStatus) {
+      // if lottery has been run before
+      throw new BadRequestException(
+        `Listing ${queryParams.listingId}: the lottery was attempted to be generated but it was already run previously`,
+      );
+    }
```

Comment:
Why would we want to prevent re-runs in the `ran` state? If Doorway doesn't approve and hasn't released they should be able to re-run

Reply:
I think that we will want another endpoint to handle "undoing" a lottery run that isn't the lottery generate endpoint

that other endpoint will be responsible for setting the lotteryStatus back to null and clearing the stored lottery positions for this listing

if we want to handle it all in the same call we could then call the lottery generate function after that work is cleaned up but I think that can be tackled a little separately

Reply:
Well wouldn't a re-run just update the stored data? I'm not sure what usecase there is for "undoing" where we set everything to null/cleared.

Reply:
I am just seeing Emily's same comment. Maybe we discuss this in huddle today

Reply:
I agree we should talk about this in the huddle. I think there is potentially a scenario where the job made it half way through and errored out leading to some applications having data and others not so we would want to start from scratch by clearing it out and starting over. But at the same time the re-run would just overwrite it so I could go either way